File: Sorting/QuickSort.java
package Sorting;

public class QuickSort {
  public static void main(String[] args) {
    Integer[] arr = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3};
    QuickSort.quickSort(arr, 0, arr.length-1);
    System.out.println(arr);
  }
  private static <T extends Comparable<T>> int partition(T[] a, int lo, int hi) {
    int i = lo, j = hi+1;
    T pivot = a[lo];
    while (true) {
      while(a[++i].compareTo(pivot) < 0) {
        if (i == hi) break;
      }
      while (pivot.compareTo(a[--j]) < 0) {
        if (j == lo) break;
      }
      if (i >= j) break;
      exch(a, i, j);
    }
    exch(a, lo, j);
    return j;
  }

  private static <T extends Comparable<T>> void exch(T[] a, int i, int j) {
    T temp = a[i];
    a[i] = a[j];
    a[j] = temp;
  }
  
  private static <T extends Comparable<T>> void quickSort(T[] a, int lo, int hi) {
    int p = partition(a, lo, hi);
    quickSort(a, lo, p-1);
    quickSort(a, p+1, hi);
  }
}

File: Sorting/QuickSort.py
class QuickSort:
  def sortArray(self, arr):
    self.arr = arr
    self.quick_sort(0, len(arr) - 1)

  def quick_sort(self, lo, hi):
    p = self.partition(lo, hi)
    self.quick_sort(lo, p-1)
    self.quick_sort(p+1, hi)

  def partition(self, lo, hi):
    pivot = self.arr[lo]
    i, j = lo+1, hi
    while True:
      print(lo, hi, self.arr)
      while i <= j and self.arr[i] < pivot:
        i += 1
      while i <= j and self.arr[j] >= pivot:
        j -= 1
      print(i, j)
      if i <= j:
        self.arr[i], self.arr[j] = self.arr[j], self.arr[i]
      else:
        break
    self.arr[lo], self.arr[j] = self.arr[j], self.arr[lo]
    return j
  

print(QuickSort().sortArray([2, 3, 1, 4]))